### PR TITLE
Skip database query when hibernate search fully satisfies search

### PR DESCRIPTION
**Pull request by michaelabuckley**
Some queries can be completely satisfied by Hibernate Search.
Before, every query was sent to the database, even when the query reduced to
```
select res_id from hfj_resource where res_id in (:res_id)
```

This change skips that redundant query when possible.
The database is still used for sorting, or when some query clauses can not be satisfied by Hibernate Search.

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/FulltextSearchSvcImpl.java [hunk 1]
```diff
@@ -271,4 +271,23 @@ private void ensureElastic() {
 		return convertLongsToResourcePersistentIds(pidList);
 	}
 
+	@Override
+	public List<IBaseResource> getResources(Collection<Long> thePids) {
+		SearchSession session = getSearchSession();
+		List<ExtendedLuceneResourceProjection> rawResourceDataList = session.search(ResourceTable.class)
+			.select(
+				f -> f.composite(
+					(pid, forcedId, resource)-> new ExtendedLuceneResourceProjection(pid, forcedId, resource),
```
Comment (jaison-b) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/FulltextSearchSvcImpl.java:280]: This is interesting. So you are basically selecting the fields and projecting it into another to do more work on it?

Comment (jaison-b) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/FulltextSearchSvcImpl.java:280]: This is interesting. So you are basically selecting the fields and projecting it into another to do more work on it?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 2]
```diff
@@ -865,23 +892,55 @@ public void loadResourcesByPid(Collection<ResourcePersistentId> thePids, Collect
...
+	 * @return can we fetch from Hibernate Search?
+	 */
+	private boolean isLoadingFromElasticSearchSupported(Collection<ResourcePersistentId> theIncludedPids) {
+		// todo mb we can be smarter here.
+		// todo check if theIncludedPids has any with version not null.
+
+		// is storage enabled?
+		return myDaoConfig.isStoreResourceInLuceneIndex() &&
+			// only support lastN for now.
+			myParams.isLastN() &&
```
Comment (jaison-b) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:919]: we only support loading from ES and skipping DB on Observation lastN operation? I was under the impression that we will support regular resource search as well(as far as we can satisfy the search params in ES).

Comment (jaison-b) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:919]: we only support loading from ES and skipping DB on Observation lastN operation? I was under the impression that we will support regular resource search as well(as far as we can satisfy the search params in ES).

**Review by jaison-b — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/FulltextSearchSvcImpl.java (hunk 1 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/FulltextSearchSvcImpl.java:280]: When you select a single field, you can provide the simple type (e.g. we were using Long before for myId).  But for multiple fields, that doesn't work.  The default return type would be `List<Object>`.  This is a bit nicer, and let's me pull out some code.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 2 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:919]: I'd like to, but this I was getting lost in all the difficulties here.  I wanted to fix the 2-second bug first.  We can come back to this later.

**Pull request merged**